components/mediators/cache/org.wso2.carbon.mediator.cache/src/main/java/org/wso2/carbon/mediator/cache/CacheMediator.java

```diff
@@ -260,86 +276,110 @@ private boolean processRequestMessage(MessageContext synCtx, SynapseLog synLog)
         cachedResponse.setResponseCodePattern(responseCodePattern);
         cachedResponse.setHTTPMethodsToCache(hTTPMethodsToCache);
         cachedResponse.setMaxMessageSize(maxMessageSize);
-        Map<String, Object> headerProperties;
+        cachedResponse.setCacheControlEnabled(cacheControlEnabled);
+        cachedResponse.setAddAgeHeaderEnabled(addAgeHeaderEnabled);
         if (cachedResponse.getResponsePayload() != null || cachedResponse.getResponseEnvelope() != null) {
             // get the response from the cache and attach to the context and change the
             // direction of the message
             if (synLog.isTraceOrDebugEnabled()) {
                 synLog.traceOrDebug("Cache-hit for message ID : " + synCtx.getMessageID());
             }
+            //Validate the response based on max-age and no-cache headers.
+            if (CachingConstants.HTTP_PROTOCOL_TYPE.equals(getProtocolType())
+                    && cachedResponse.isCacheControlEnabled() &&
+                    HttpCachingFilter.isValidCacheEntry(cachedResponse, synCtx)) {
+                return true;
+            }
             // mark as a response and replace envelope from cache
             synCtx.setResponse(true);
-            try {
-                if (cachedResponse.isJson()) {
-                    byte[] payload = cachedResponse.getResponsePayload();
-                    OMElement response = JsonUtil.getNewJsonPayload(msgCtx, payload, 0,
-                                                                    payload.length, false, false);
-                    if (msgCtx.getEnvelope().getBody().getFirstElement() != null) {
-                        msgCtx.getEnvelope().getBody().getFirstElement().detach();
-                    }
-                    msgCtx.getEnvelope().getBody().addChild(response);
+            returnCachedResponse(synCtx, synLog, msgCtx, cachedResponse);
+            return false;
+        }
+        return true;
+    }
 
-                } else {
-                    msgCtx.setEnvelope(MessageHelper.cloneSOAPEnvelope(cachedResponse.getResponseEnvelope()));
+    /**
+     * This method returns the existing cached response.
+     * @param synCtx Message context.
+     * @param synLog Synapse log.
+     * @param msgCtx Axis2 contex.
+     * @param cachedResponse Cached response.
+     */
+    private void returnCachedResponse(MessageContext synCtx, SynapseLog synLog,
+                                      org.apache.axis2.context.MessageContext msgCtx, CachableResponse cachedResponse) {
```

Comment:
This name is very strange. The method does not return anything and the keyword return is used. We normally use "get" when a method is returning anything.